File: backend/app/media.py
"""Media surfacing for the chat stream.

The deepagents agent exposes media two ways:
  1. It WRITES files into the per-session workspace (served by /v1/workspace/raw)
     and the LLM may reference an absolute path with a ``MEDIA:<path>`` line (a
     prompt convention; present in imported hermes conversations).
  2. When it READS a binary file, ``read_file`` returns a ToolMessage whose
     content is LangChain multimodal blocks ``{"type":"image"|"audio"|"video"|
     "file", "base64":..., "mime_type":...}`` (see deepagents
     middleware/filesystem.py) — otherwise flattened away by ``_content_to_text``.

``resolve_media`` safely serves an absolute local file (1); ``media_from_message``
extracts base64 blocks into stream-friendly media descriptors (2).
"""

from __future__ import annotations
import asyncio
import hashlib
import logging
import mimetypes
import os
import re
import shutil
import subprocess
import tempfile

from .constants import DEFAULT_USER_ID, MAX_MEDIA_B64_BYTES, MAX_MEDIA_BYTES, OFFICE_TO_PDF_TIMEOUT_S

logger = logging.getLogger("joyjoy.media")

# Extension groups we serve / preview. Images, A/V and PDF render directly; office
# docs are converted to PDF on demand (office_to_pdf); text/code/markdown files are
# fetched + rendered by the client. Markers the LLM emits can point at any of these.
IMAGE_EXTS = {
    ".png", ".jpg", ".jpeg", ".gif", ".webp", ".bmp", ".ico", ".svg", ".heic", ".heif",
}
PDF_EXTS = {".pdf"}
OFFICE_EXTS = {
    ".doc", ".docx", ".xls", ".xlsx", ".ppt", ".pptx",
    ".odt", ".ods", ".odp", ".rtf",
}
AV_EXTS = {".mp3", ".wav", ".ogg", ".m4a", ".aac", ".flac", ".mp4", ".webm", ".mov", ".m4v"}
TEXT_EXTS = {
    ".md", ".markdown", ".txt", ".csv", ".tsv", ".json", ".jsonl", ".xml",
    ".yaml", ".yml", ".html", ".htm", ".css", ".py", ".js", ".ts", ".tsx",
    ".jsx", ".mjs", ".cjs", ".sh", ".bash", ".zsh", ".go", ".rs", ".java",
    ".kt", ".c", ".cc", ".cpp", ".h", ".hpp", ".rb", ".php", ".sql", ".toml",
    ".ini", ".cfg", ".conf", ".log",
}
ALLOWED_EXTS = IMAGE_EXTS | PDF_EXTS | OFFICE_EXTS | AV_EXTS | TEXT_EXTS

# Byte caps centralized in constants.py (MAX_MEDIA_BYTES / MAX_MEDIA_B64_BYTES).


def is_office(path: str) -> bool:
    return os.path.splitext(path)[1].lower() in OFFICE_EXTS


# Cache of converted PDFs, keyed by (source path, mtime) so re-views are instant.
_PDF_CACHE = os.path.join(tempfile.gettempdir(), "joyjoy_office_pdf")
_SOFFICE = shutil.which("soffice") or shutil.which("libreoffice")


def _convert_office_to_pdf(src: str) -> str | None:
    """Convert an office doc to PDF via headless LibreOffice; cache by mtime.
    Returns the PDF path, or None if soffice is unavailable or conversion fails."""
    if not _SOFFICE or not os.path.isfile(src):
        return None
    try:
        mtime = os.path.getmtime(src)
    except OSError:
        return None
    key = hashlib.sha1(f"{os.path.realpath(src)}:{mtime}".encode()).hexdigest()
    os.makedirs(_PDF_CACHE, exist_ok=True)
    out_pdf = os.path.join(_PDF_CACHE, f"{key}.pdf")
    if os.path.isfile(out_pdf):
        return out_pdf
    # A per-conversion LibreOffice profile dir avoids the "already running" lock
    # when several conversions overlap.
    profile = os.path.join(_PDF_CACHE, f"profile-{key}")
    try:
        subprocess.run(
            [
                _SOFFICE, "--headless", "--norestore",
                f"-env:UserInstallation=file://{profile}",
                "--convert-to", "pdf", "--outdir", _PDF_CACHE, src,
            ],
            capture_output=True, timeout=OFFICE_TO_PDF_TIMEOUT_S, check=False,
        )
    except Exception:  # noqa: BLE001 - treat any failure as "no preview"
        logger.warning("office→pdf conversion failed for %s", src, exc_info=True)
        return None
    finally:
        shutil.rmtree(profile, ignore_errors=True)
    produced = os.path.join(_PDF_CACHE, os.path.splitext(os.path.basename(src))[0] + ".pdf")
    if os.path.isfile(produced):
        if produced != out_pdf:
            try:
                os.replace(produced, out_pdf)
            except OSError:
                return produced
        return out_pdf
    return None


async def office_to_pdf(src: str) -> str | None:
    """Async wrapper — runs the (blocking) LibreOffice conversion off the event loop."""
    return await asyncio.to_thread(_convert_office_to_pdf, src)


def _win_to_wsl(p: str) -> str:
    """Translate a Windows path (``C:\\Users\\…`` — common in imported hermes
    conversations) to its WSL mount (``/mnt/c/Users/…``)."""
    m = re.match(r"^([A-Za-z]):[\\/](.*)$", p)
    if not m:
        return p
    drive, rest = m.group(1).lower(), m.group(2).replace("\\", "/")
    return f"/mnt/{drive}/{rest}"


def _safe_roots(settings, user_id: str) -> list[str]:
    """Dirs an absolute media path is allowed to live under (realpath'd).

    Prod is locked to the per-user workspace only. Extra host roots (configured via
    ``MEDIA_DEV_EXTRA_ROOTS``, e.g. the WSL home or Windows ``/mnt/c/Users``) are
    DEV-only — a convenience for testing imported-conversation media that references
    absolute host paths. In prod they'd let a crafted ``MEDIA:`` marker surface
    arbitrary host/home files to the chat client, so they're ignored there.
    """
    # Use the SAME root as the workspace dock/agent (WORKSPACE_ROOT, falling back to
    # user_data_root) so host-mode MEDIA: markers resolve where files actually live.
    ws = os.path.join(settings.workspace_root_dir, str(user_id or DEFAULT_USER_ID), "workspace")
    cands = [ws]
    if not settings.is_prod:
        cands += settings.media_dev_extra_root_list
    roots: list[str] = []
    for cand in cands:
        try:
            rp = os.path.realpath(cand)
            if os.path.isdir(rp):
                roots.append(rp)
        except OSError:
            continue
    return roots


def resolve_media(settings, user_id: str, raw_path: str) -> tuple[str, str] | None:
    """(absolute path, mime) for a serveable local media file, or None if missing,
    too large, the wrong type, or outside the allowed roots."""
    if not raw_path:
        return None
    cleaned = _win_to_wsl(raw_path.strip().strip("\"'`"))
    full = os.path.realpath(os.path.expanduser(cleaned))
    if os.path.splitext(full)[1].lower() not in ALLOWED_EXTS:
        return None
    roots = _safe_roots(settings, user_id)
    if not any(full == r or full.startswith(r + os.sep) for r in roots):
        return None
    if not os.path.isfile(full):
        return None
    try:
        if os.path.getsize(full) > MAX_MEDIA_BYTES:
            return None
    except OSError:
        return None
    mime = mimetypes.guess_type(full)[0] or "application/octet-stream"
    return full, mime


def media_from_message(m) -> list[dict]:
    """Extract base64 media blocks from a (Tool)Message into stream descriptors:
    ``{kind, mime_type, filename?, data_url}``. Empty for plain-text messages."""
    blocks = None
    content = getattr(m, "content", None)
    if isinstance(content, list):
        blocks = content
    else:
        cb = getattr(m, "content_blocks", None)
        if isinstance(cb, list):
            blocks = cb
    if not blocks:
        return []

    ak = getattr(m, "additional_kwargs", None) or {}
    src = ak.get("read_file_path")
    filename = os.path.basename(src) if isinstance(src, str) and src else None

    out: list[dict] = []
    for b in blocks:
        if not isinstance(b, dict):
            continue
        kind = b.get("type")
        if kind not in ("image", "audio", "video", "file"):
            continue
        b64 = b.get("base64") or b.get("data")
        if not isinstance(b64, str) or not b64 or len(b64) > MAX_MEDIA_B64_BYTES:
            continue
        mime = b.get("mime_type") or b.get("mimeType") or "application/octet-stream"
        out.append(
            {
                "kind": kind,
                "mime_type": mime,
                "filename": filename,
                "data_url": f"data:{mime};base64,{b64}",
            }
        )
    return out

File: frontend/src/lib/media.ts
// Media helpers for chat rendering. Agent media comes from: absolute paths via a
// `MEDIA:<path>` line (→ /v1/media?path=), files the agent wrote into the
// per-session workspace (→ /v1/workspace/raw?thread_id&path), and base64 blocks.
// Office docs are converted to PDF server-side (?convert=pdf); text/code/markdown
// files are fetched and rendered client-side. All go through the /v1 dev proxy.

const IMAGE_EXTS = new Set([
	".png",
	".jpg",
	".jpeg",
	".gif",
	".webp",
	".bmp",
	".ico",
	".svg",
	".heic",
	".heif",
	".avif",
]);
const PDF_EXTS = new Set([".pdf"]);
const OFFICE_EXTS = new Set([
	".doc",
	".docx",
	".xls",
	".xlsx",
	".ppt",
	".pptx",
	".odt",
	".ods",
	".odp",
	".rtf",
]);
const AUDIO_EXTS = new Set([".mp3", ".wav", ".ogg", ".m4a", ".aac", ".flac"]);
const VIDEO_EXTS = new Set([".mp4", ".webm", ".mov", ".m4v"]);
const AV_EXTS = new Set([...AUDIO_EXTS, ...VIDEO_EXTS]);
const MARKDOWN_EXTS = new Set([".md", ".markdown"]);
const TEXT_EXTS = new Set([
	...MARKDOWN_EXTS,
	".txt",
	".csv",
	".tsv",
	".json",
	".jsonl",
	".xml",
	".yaml",
	".yml",
	".html",
	".htm",
	".css",
	".py",
	".js",
	".ts",
	".tsx",
	".jsx",
	".mjs",
	".cjs",
	".sh",
	".bash",
	".zsh",
	".go",
	".rs",
	".java",
	".kt",
	".c",
	".cc",
	".cpp",
	".h",
	".hpp",
	".rb",
	".php",
	".sql",
	".toml",
	".ini",
	".cfg",
	".conf",
	".log",
]);
// "Media" = binary/visual artifacts worth auto-previewing when the agent WRITES
// one (write_file). Text/code/markdown are excluded here — they only render inline
// when the agent explicitly emits a MEDIA: marker (else every code file it writes
// would clutter the chat).
const MEDIA_EXTS = new Set([
	...IMAGE_EXTS,
	...PDF_EXTS,
	...OFFICE_EXTS,
	...AV_EXTS,
]);

const MIME: Record<string, string> = {
	".png": "image/png",
	".jpg": "image/jpeg",
	".jpeg": "image/jpeg",
	".gif": "image/gif",
	".webp": "image/webp",
	".svg": "image/svg+xml",
	".bmp": "image/bmp",
	".ico": "image/x-icon",
	".pdf": "application/pdf",
	".mp3": "audio/mpeg",
	".wav": "audio/wav",
	".ogg": "audio/ogg",
	".mp4": "video/mp4",
	".webm": "video/webm",
	".mov": "video/quicktime",
	".docx":
		"application/vnd.openxmlformats-officedocument.wordprocessingml.document",
	".xlsx": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
	".pptx":
		"application/vnd.openxmlformats-officedocument.presentationml.presentation",
	".doc": "application/msword",
	".xls": "application/vnd.ms-excel",
	".ppt": "application/vnd.ms-powerpoint",
	".md": "text/markdown",
	".txt": "text/plain",
	".csv": "text/csv",
	".json": "application/json",
};

export const extOf = (name: string): string => {
	const i = name.lastIndexOf(".");
	return i >= 0 ? name.slice(i).toLowerCase() : "";
};
// Human-readable byte size (B / KB / MB). Empty string for null/undefined.
export const formatSize = (bytes?: number): string => {
	if (bytes == null) return "";
	if (bytes < 1024) return `${bytes} B`;
	if (bytes < 1024 * 1024) return `${(bytes / 1024).toFixed(1)} KB`;
	return `${(bytes / (1024 * 1024)).toFixed(1)} MB`;
};
export const isImageFile = (name: string): boolean =>
	IMAGE_EXTS.has(extOf(name));
export const isAudioFile = (name: string): boolean =>
	AUDIO_EXTS.has(extOf(name));
export const isVideoFile = (name: string): boolean =>
	VIDEO_EXTS.has(extOf(name));
export const isPdfFile = (name: string): boolean => PDF_EXTS.has(extOf(name));
export const isOfficeFile = (name: string): boolean =>
	OFFICE_EXTS.has(extOf(name));
export const isTextFile = (name: string): boolean => TEXT_EXTS.has(extOf(name));
export const isMarkdownFile = (name: string): boolean =>
	MARKDOWN_EXTS.has(extOf(name));
// Auto-previewable media artifact (used to gate write_file previews).
export const isMediaFile = (name: string): boolean =>
	MEDIA_EXTS.has(extOf(name));
export const mimeOf = (name: string): string =>
	MIME[extOf(name)] ?? "application/octet-stream";

// True for an inline `data:` URL (already-embedded bytes — no fetch needed).
export const isDataUrl = (url: string): boolean => url.startsWith("data:");

// thread_id lets the backend resolve the marker inside the session's sandbox
// volume when SANDBOX_ENABLED (else it's ignored and a host path is used).
export const mediaUrl = (threadId: string, path: string): string =>
	`/v1/media?thread_id=${encodeURIComponent(threadId)}&path=${encodeURIComponent(path)}`;
export const workspaceRawUrl = (threadId: string, path: string): string =>
	`/v1/workspace/raw?thread_id=${encodeURIComponent(threadId)}&path=${encodeURIComponent(path)}`;
// Ask the server to convert an office doc to PDF for inline preview. Both source
// URLs already carry a query string, so always append with `&`.
export const asPdf = (url: string): string => `${url}&convert=pdf`;

// Whole-line `MEDIA:<path>` marker (hermes convention). Tolerates wrapping
// quotes/backticks and paths with spaces; captures the path.
const MEDIA_LINE_RE = /^[ \t]*[`"']?MEDIA:[ \t]*(.+?)[`"']?[ \t]*$/;

export type TextSegment = { kind: "text"; text: string };
export type MediaSegment = { kind: "media"; path: string };

// Split assistant text into ordered text/media segments, lifting MEDIA: lines out.
export function splitMediaMarkers(
	text: string,
): (TextSegment | MediaSegment)[] {
	if (!text.includes("MEDIA:")) return text ? [{ kind: "text", text }] : [];
	const segs: (TextSegment | MediaSegment)[] = [];
	let buf: string[] = [];
	const flush = () => {
		const joined = buf.join("\n");
		if (joined.trim()) segs.push({ kind: "text", text: joined });
		buf = [];
	};
	for (const line of text.split("\n")) {
		const m = line.match(MEDIA_LINE_RE);
		if (m) {
			flush();
			segs.push({ kind: "media", path: m[1].trim() });
		} else {
			buf.push(line);
		}
	}
	flush();
	return segs;
}
